### ISSUE TITLE
EOF must be sent twice on stdin if no other input is sent

### ISSUE THREAD

**Issue by willcassella** [sharkdp/bat#477]
If you run `bat` and then immediately press the EOF sequence (Ctrl-D), bat prints the filename 'STDIN' but continues to read from it. You have to press it again to actually close the file.

**Comment by sharkdp** [comment]
Hm, interesting. Thank you very much for reporting this.

**Comment by mziter** [comment]
I have found and resolved issue locally. However, I am not sure how to exit immeditately as process::exit(0) will exit immediately and not do any destructuring, etc...

Other two options are 1) Wrap new function in a result, which seems idiomatic, but would require a lot of downstream changes I am sure. 2) Wrap in a new ContentType that would basically get passed on and ignored at some future point.

Would like your thoughts...

**Comment by mziter** [comment]
Sorry, just saw this referenced already. :) Will put some comments on pull request.

**Comment by sharkdp** [comment]
closed via #492 by @reidwagner

**Comment by sharkdp** [comment]
Fixed in [v0.10.0](https://github.com/sharkdp/bat/releases/tag/v0.10.0).

**Comment by reidwagner** [comment]
This has been closed - but there was a change in expected behavior, where we no longer get a header and footer for empty input files.

Trying to bring back that behavior while keeping the error logic here seemed to get complex. I think I have a simpler solution all around.

Instead of what's implemented in #492 (that could be reverted), we simply don't call `print_file_ranges()` if our reader.first_line is empty. That appears to give all the desired functionality, + it prints the STDIN header and footer when we press CTRL-D, which seems more consistent in the first place. e.g in `print_file()`:       
```
if !reader.first_line.is_empty() {
    self.print_file_ranges(printer, writer, reader, &self.config.line_ranges)?;
}
```

I _think_ maybe the check that `first_line.is_empty()` could also be removed from `read_line` since the condition above would work for empty files.

**Comment by sharkdp** [comment]
> Instead of what's implemented in #492 (that could be reverted), we simply don't call `print_file_ranges()` if our reader.first_line is empty. That appears to give all the desired functionality, + it prints the STDIN header and footer when we press CTRL-D, which seems more consistent in the first place

Sounds great!

> I _think_ maybe the check that `first_line.is_empty()` could also be removed from `read_line` since the condition above would work for empty files.

I can't tell, at the moment. We have to make sure to correctly handle (1) empty files (2) files with a single line (3) files with a single line and no `\n` (4) files with multiple lines (6) files with other encodings (mainly UTF-16 for now). I hope that most of this should be captured by tests, but it might be worth to double-check.

**Comment by reidwagner** [comment]
So I opened #504 with the simple change proposed + reverting #492, in case you want to get the fix in soon. As for removing the code from read_line I can double check that before making the change (either added to #504 or another PR).
